docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,343 @@
+
+  Also, please note that while we would like to align with Kubernetes and leverage its new features as soon as 
+  possible, there will always be a delta to be addressed in Cluster API due to the fact that Cluster API is operated 
+  from the management cluster, while drain is done in workload clusters.
+
+  Additionally, Cluster API should take care of many clusters, while Kubernetes features are by design scoped to a single cluster.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### `MachineDrainRule` CRD
+
+We propose to introduce a new `MachineDrainRule` namespace-scoped CRD. The goal of this CRD is to allow management 
+cluster admins/users to define a drain rule (a specific drain behavior) for a set of Pods.
+
+Additionally, drain rules can be restricted to a specific set of Machines, e.g. targeting only controlplane Machines, 
+Machines of specific Clusters, Linux Machines, etc.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: example-rule
+  namespace: default
+spec:
+  drain:
+    # behavior can be Drain or Skip
+    behavior: Drain 
+    # order can only be set if behavior == Drain
+    # Pods with higher order are drained later
+    order: 100
+  machines:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    clusterSelector:
+      matchLabels:
+      matchExpressions:
+  pods:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    namespaceSelector:
+      matchLabels:
+      matchExpressions:
+```
+
+Notes:
+
+* All selectors are of type `metav1.LabelSelector`
+* In order to allow expressing selectors with AND and OR predicates, both `machines` and `pods` have a list of 
+  selectors instead of a single selector
+
+##### Example: Exclude Pods from drain
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: skip-pods
+  namespace: default
+spec:
+  drain:
+    behavior: Skip
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchExpressions:
+      - key: app
+        operator: In
+        values:
+        - example-app1
+        - example-app2
+        - example-app3
+    namespaceSelector:
+      matchLabels:
+        kubernetes.io/metadata.name: example-namespace
+```
+
+##### Example: Drain order
+
+(related to [CAPI-11024](https://github.com/kubernetes-sigs/cluster-api/issues/11024))
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: portworx
+  namespace: default
+spec:
+  drain:
+    behavior: Drain
+    order: 100
+  machines:
+  - selector: {}
+  pods:
+  - selector:
+      matchLabels:
+        app: portworx
+```
+
+#### Drain labels
+
+We propose to introduce new labels to allow workload cluster admins/users to define drain behavior 
+for Pods. These labels would be either added directly to Pods or indirectly via Deployments, StatefulSets, etc.
+The labels will take precedence over `MachineDrainRules` specified in the management cluster.
```

Comment:
What if I'm a fleet admin, and I don't want to allow the workload admins to be able to override the drain policy I've implemented? Do I have a get out of jail free card?

What issues are we preventing by allowing workload admins to overrule the fleet admins?

Reply:
The idea is that the workload admin owns the workload. So they should be able to decide if a workload should not be drained.

What is the use case for that a fleet admin enforces that a specific workload has to be drained?

Reply:
> What is the use case for that a fleet admin enforces that a specific workload has to be drained? (I.e. it cannot be skipped)

I think this comes back to my concern about volumes. What if a workload admin doesn't realise that their workload skipping draining is going to cause the fleet admin issues, because of side effects of that pod still being present on the node?

I guess the drain policy isn't really scoped enough to be able to detect specific pod properties like, has a volume attached, so, this probably doesn't help specifically in this case anyway, it would have to be label based (though through policy they could enforce a specific label 🤔)

Reply:
The more we can make APIs declarative, the more we can capture that people / principals can have conflicting intents.

eg
ReplicaSet: I intend to make 2000 Pods
ResourceQuota: this namespace can have at most 42 Pods

eg
MachineDrainRules: This node / these nodes are going offline by 1930 UTC, today (a weekday)
workload disruption budget: I can only have at most one replica offline at once, nothing more.
StatefulSet: these Pods have an identity from ordinal 7 through 9
norkload operating hours: we need this disruption budget in place every weekday even public holidays
node autoscaling: We need to have at least 16vCPU, 3GiB of RAM available as overhead capacity

eg
storage: this storage can only be in zone antartica-1b
workload: these pods can be in any zone in Antarctica
machines: nodes in zone antarctica-1b are going down for maintenance
resilience: create replacement pods before planned disruptions, mkay?